lib/reversetunnel/agent.go

```diff
@@ -22,511 +22,557 @@ package reversetunnel
+	a.log.Debugf("Changing state %s -> %s.", prevState, state)
+
+	if a.agentConfig.stateCallback != nil {
+		go a.agentConfig.stateCallback(a.state)
+	}
+
+	return prevState, nil
 }
 
-func (a *Agent) getHostCheckers() ([]ssh.PublicKey, error) {
-	cas, err := a.AccessPoint.GetCertAuthorities(context.TODO(), types.HostCA, false)
+// Start starts an agent returning after successfully connecting and sending
+// the first heatbeat.
+func (a *agent) Start(ctx context.Context) error {
+	a.log.Debugf("Starting agent %v", a.addr)
+
+	var err error
+	defer func() {
+		a.once.Do(func() {
+			close(a.doneConnecting)
+		})
+		if err != nil {
+			a.Stop()
+		}
+	}()
+
+	_, err = a.updateState(AgentConnecting)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
-	var keys []ssh.PublicKey
-	for _, ca := range cas {
-		checkers, err := sshutils.GetCheckers(ca)
-		if err != nil {
-			return nil, trace.Wrap(err)
+
+	a.ctx, a.cancel = context.WithCancel(ctx)
+	a.drainCtx, a.drainCancel = context.WithCancel(a.ctx)
+
+	err = a.connect()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Start handing global requests again.
+	a.wg.Add(1)
+	go func() {
+		if err := a.handleGlobalRequests(a.ctx, a.client.GlobalRequests()); err != nil {
+			a.log.WithError(err).Debug("Failed to handle global requests.")
+		}
+		a.wg.Done()
+		a.Stop()
+	}()
+
+	// drainWG.Done will be called from handleDrainChannels.
+	a.drainWG.Add(1)
+	a.wg.Add(1)
+	go func() {
+		if err := a.handleDrainChannels(); err != nil {
+			a.log.WithError(err).Debug("Failed to handle drainable channels.")
 		}
-		keys = append(keys, checkers...)
+		a.wg.Done()
+		a.Stop()
+	}()
+
+	a.wg.Add(1)
+	go func() {
+		if err := a.handleChannels(); err != nil {
+			a.log.WithError(err).Debug("Failed to handle channels.")
+		}
+		a.wg.Done()
+		a.Stop()
+	}()
+
+	_, err = a.updateState(AgentConnected)
+	if err != nil {
+		return trace.Wrap(err)
 	}
-	return keys, nil
+
+	return nil
 }
 
-// getReverseTunnelDetails pings the remote Teleport Proxy address in order to check if this is Web Service or ReverseTunnel Service address.
-// If this is Web Service port check if proxy support ALPN SNI Listener.
-func (a *Agent) getReverseTunnelDetails() *reverseTunnelDetails {
-	pd := reverseTunnelDetails{TLSRoutingEnabled: false}
-	resp, err := webclient.Find(
-		&webclient.Config{Context: a.ctx, ProxyAddr: a.Addr.Addr, Insecure: lib.IsInsecureDevMode(), IgnoreHTTPProxy: true})
+// connect connects to the server and finishes setting up the agent.
+func (a *agent) connect() error {
+	client, err := a.sshDialer.DialContext(a.ctx, a.addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	a.client = client
+
+	unclaim, ok := a.tracker.Claim(a.client.Principals()...)
+	if !ok {
+		a.client.Close()
+		return trace.Errorf("Failed to claim proxy: %v claimed by another agent", a.client.Principals())
+	}
+	a.unclaim = unclaim
 
+	startupCtx, cancel := context.WithCancel(a.ctx)
+	done := make(chan struct{})
```

Comment:
This `done` channel can be removed, and the later defer can just be a `defer cancel()` right here.

Reply:
This wouldn't guarantee `handleGlobalRequests` returns before we return from `connect` but I think that should be fine.

Reply:
Updated here https://github.com/gravitational/teleport/pull/12359/commits/b360af03f0008f5a8a2c2442435a87b84a816e09